plugins/database/src/general_query.cpp

```diff
@@ -1164,56 +1167,102 @@ checkCondition( char *condition ) {
     return 0;
 }
 
-
 /*
 add an IN clause to the whereSQL string for the Parent_Of option
  */
 int
-addInClauseToWhereForParentOf( char *inArg ) {
-    int i, len;
-    int nput = 0;
-    char tmpStr[MAX_SQL_SIZE_GQ];
-
-    // [#3913] silent failure of imkdir -p with long path containing white space
-    static char inStrings[MAX_SQL_SIZE_GQ*2];
-    int inStrIx = 0;
-
-    if ( !rstrcat( whereSQL, " IN (", MAX_SQL_SIZE_GQ ) ) { return USER_STRLEN_TOOLONG; }
-    len = strlen( inArg );
-    for ( i = 0; i < len + 1; i++ ) {
-        if ( inArg[i] == '/' || inArg[i] == '\0' ) {
-            int ncopy = i;
-            if ( nput == 0 ) {
-                ncopy++;
-            }
-            if ( nput == 0 ) {
-                if ( !rstrcat( whereSQL, "?", MAX_SQL_SIZE_GQ ) ) { return USER_STRLEN_TOOLONG; }
-            }
-            else {
-                if ( !rstrcat( whereSQL, ", ?", MAX_SQL_SIZE_GQ ) ) { return USER_STRLEN_TOOLONG; }
+addInClauseToWhereForParentOf( char *inArg )
+{
+    // This vector holds all of the components of the path
+    // in the parameter inArg, with no slashes.
+    std::vector<std::string> paths;
+
+    // The purpose of this vector of strings is to stick around until all
+    // references to the strings included are gone -- the cllBindVars[] array
+    // of pointers to null terminated strings is filled with these strings
+    // which cannot be removed until the next call to this function.
+    static std::vector<std::string> inStringVec;
+    static bool reset_vector = false;
+
+    try {
+        std::string stringInArg(const_cast<const char *>(inArg));
+        boost::algorithm::split(paths, stringInArg, boost::is_any_of(irods::get_virtual_path_separator()));
+    }
+    catch ( const boost::bad_function_call& ) {
+        rodsLog( LOG_ERROR, "boost::split threw boost::bad_function_call" );
+        return BAD_FUNCTION_CALL;
+    }
+
+    // Collect all the parameter path components in order
+    // from left to right.  Starting from the second call to
+    // this function, the vector is cleared as explained above.
+    if (reset_vector)
+    {
+        inStringVec.clear();
+    }
+    reset_vector = true;
+
+    // Put together all the paths included in the parameter path
+    // and save them in the static vector.  These strings will
+    // be assigned to the global bind variable array used
+    // in the WHERE clause.
+    //
+    // Thusly, the path "/tempZone/trash/home/public" for example, will become:
+    //
+    //           inStringVec[0] = /
+    //           inStringVec[1] = /tempZone
+    //           inStringVec[2] = /tempZone/trash
+    //           inStringVec[3] = /tempZone/trash/home
+    //           inStringVec[4] = /tempZone/trash/home/public
+    for (size_t si = 0; si < paths.size(); si++)
+    {
+        std::string path;
+        bool need_slash = false;
+
+        for (size_t j = 0; j <= si; j++)
+        {
+            if (paths[j].empty()) {
+                path += "/";
```

Comment:
Both here and in the next few lines, the hard-coded "/" will be changed to irods::get_virtual_path_separator() before this code can be merged.